master/jobmaster/benchmark/master.go

```diff
@@ -0,0 +1,141 @@
+package benchmark
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/hanfei1991/microcosom/master/cluster"
+	"github.com/hanfei1991/microcosom/model"
+	"github.com/hanfei1991/microcosom/pb"
+)
+
+type Master struct {
+	*Config
+	job *model.Job
+
+	resouceManager cluster.ResourceMgr
+	client         cluster.ExecutorClient
+}
+
+var lagThredshold int = 10
+
+type TaskStatus int
+
+const (
+	Initing TaskStatus = iota
+	Running
+	Disconnect
+	Dead
+	Finished
+)
+
+type Task struct {
+	sync.Mutex
+	*model.Task
+
+	lag              int
+	lastScheduleTime time.Time
+}
+
+func (m *Master) ID() model.JobID {
+	return m.job.ID
+}
+
+func (m *Master) dispatch(tasks []*model.Task) error {
+	arrangement := make(map[model.ExecutorID][]*model.Task)
+	for _, task := range tasks {
+		subjob, ok := arrangement[task.ExecutorID]
+		if !ok {
+			arrangement[task.ExecutorID] = []*model.Task{task}
+		} else {
+			subjob = append(subjob, task)
+		}
+	}
+	// TODO: process the error cases.
+	for execID, taskList := range arrangement {
+		// construct sub job
+		job := &model.Job{
+			ID: m.job.ID,
+			Tasks: taskList,
+		}
+		reqPb := job.ToPB()
+		request := &cluster.ExecutorRequest{
+			Cmd: cluster.CmdSubmitSubJob,
+			Req: reqPb,
+		}
+		resp, err := m.client.Send(context.Background(), execID, request)
+		if err != nil {
+			return err
+		}
+		respPb := resp.Resp.(*pb.SubmitJobResponse)
+		if respPb.ErrMessage != "" {
+			return errors.New(respPb.ErrMessage)
+		}
+	}
+	return nil
+}
+
+// TODO: Implement different allocate task logic.
+func (m *Master) allocateTasksWithNaitiveStratgy(snapshot *cluster.ResourceSnapshot) (bool, []*model.Task) {
+	var idx int = 0
+	for _, task := range m.job.Tasks {
+		originalIdx := idx
+		for {
```

Comment:
Implementing in the job master leaves more space to adjust the allocation algo. Say that a set of task need to be allocated in same node, some task have locality peference. Different types of Job might have different strategy of allocation.

Reply:
We can use a allocation algo helper to implement the allocing strategy. It's set to refine in the future.
It can be decided by the types of sub-plan.